src/sys/socket/mod.rs

```diff
@@ -392,53 +403,25 @@ impl<'a> Iterator for CmsgIterator<'a> {
     // although we handle the invariants in slightly different places to
     // get a better iterator interface.
     fn next(&mut self) -> Option<ControlMessage<'a>> {
-        let sizeof_cmsghdr = mem::size_of::<cmsghdr>();
-        if self.buf.len() < sizeof_cmsghdr {
+        if self.buf.len() == 0 {
+            // The iterator assumes that `self.buf` always contains exactly the
+            // bytes we need, so we're at the end when the buffer is empty.
             return None;
         }
-        let cmsg: &'a cmsghdr = unsafe { &*(self.buf.as_ptr() as *const cmsghdr) };
 
-        // This check is only in the glibc implementation of CMSG_NXTHDR
-        // (although it claims the kernel header checks this), but such
-        // a structure is clearly invalid, either way.
+        // Safe if: `self.buf` is `cmsghdr`-aligned.
+        let cmsg: &'a cmsghdr = unsafe { &*(self.buf[..mem::size_of::<cmsghdr>()].as_ptr() as *const cmsghdr) };
+
         let cmsg_len = cmsg.cmsg_len as usize;
-        if cmsg_len < sizeof_cmsghdr {
-            return None;
-        }
-        let len = cmsg_len - sizeof_cmsghdr;
-        let aligned_cmsg_len = if self.next == 0 {
-            // CMSG_FIRSTHDR
-            cmsg_len
-        } else {
-            // CMSG_NXTHDR
-            cmsg_align(cmsg_len)
-        };
 
         // Advance our internal pointer.
-        if aligned_cmsg_len > self.buf.len() {
-            return None;
-        }
-        let cmsg_data = &self.buf[cmsg_align(sizeof_cmsghdr)..cmsg_len];
-        self.buf = &self.buf[aligned_cmsg_len..];
-        self.next += 1;
-
-        match (cmsg.cmsg_level, cmsg.cmsg_type) {
-            (libc::SOL_SOCKET, libc::SCM_RIGHTS) => unsafe {
-                Some(ControlMessage::ScmRights(
-                    slice::from_raw_parts(cmsg_data.as_ptr() as *const _,
-                                          cmsg_data.len() / mem::size_of::<RawFd>())))
-            },
-            (libc::SOL_SOCKET, libc::SCM_TIMESTAMP) => unsafe {
-                Some(ControlMessage::ScmTimestamp(
-                    &*(cmsg_data.as_ptr() as *const _)))
-            },
-            (_, _) => unsafe {
-                Some(ControlMessage::Unknown(UnknownCmsg(
-                    cmsg,
-                    slice::from_raw_parts(
-                        cmsg_data.as_ptr() as *const _,
-                        len))))
-            }
+        let cmsg_data = &self.buf[cmsg_align(mem::size_of::<cmsghdr>())..cmsg_len];
```

Comment:
Up above you requird self.buf to be aligned.  Here, you seem to be handling the case where it isn't.  Or am I misunderstanding?

Reply:
The `cmsg_align` just rounds up the size of `cmsghdr` to the right alignment. This is needed to include the right padding to reach the payload.